Flow/v1

cryogenian opened the pull request:
[ch13061]

+ Removed newtypes in config
+ Ingesting per chunk, not whole stream
+ `Flow` implemented
+ Tests are kind of implemented, there is no destination types here and most of things are meaningless. 
+ No `autocommit` 

`includeHeader = false` and `skip_headers = 0` are a bit strange, because w/o `includeHeader = false` there is no headers anyways.

rintcius commented on core/src/test/scala/quasar/destination/snowflake/SnowflakeDestinationSpec.scala:
```diff
@@ -0,0 +1,791 @@
+  def User = sys.env.get("SNOWFLAKE_USER") getOrElse ""
+  def Database = sys.env.get("SNOWFLAKE_DATABASE") getOrElse ""
+  def Schema = sys.env.get("SNOWFLAKE_SCHEMA") getOrElse ""
+  def Warehouse = sys.env.get("SNOWFLAKE_WAREHOUSE") getOrElse ""
```
Possible to run this spec in CI if we define these as repos secrets and inject them in build.yml?

cryogenian replied:
In theory, might cost something though. At least we can test it locally

wemrysi replied:
To confirm, have these all been successfully run locally?

cryogenian replied:
Yes, tests pass locally

rintcius reviewed (commented):
Looks great! Just a couple of comments.

cryogenian commented on core/src/main/scala/quasar/destination/snowflake/SnowflakeConfig.scala:
```diff
@@ -21,48 +21,50 @@ import scala._
-          cfg.warehouse.value,
-          cfg.sanitizeIdentifiers.value.some).some
-    )("accountName", "user", "password", "databaseName", "schema", "warehouse", "sanitizeIdentifiers")
+  val Redacted = "<REDACTED>"
```
MariaDB uses jdbc uri, iiuc this is the reason for double dash instead of lt/gt.

This is goid old config, only jsons object presented as a form

rintcius replied:
Ah yes, you are right - not relevant here

cryogenian commented on core/src/main/scala/quasar/destination/snowflake/QueryGen.scala:
```diff
@@ -22,8 +22,8 @@ import scala.util.matching.Regex
   // we should eventually move the query generation code here to make it more testable
-  def sanitizeIdentifier(str: String, cfg: SanitizeIdentifiers): String =
-    if (cfg.value)
+  def sanitizeIdentifier(str: String, cfg: Boolean): String =
```
Hm... `toUpperCase`?

cryogenian replied:
Or `snowflakeSanitation`?

wemrysi commented on core/src/main/scala/quasar/destination/snowflake/SnowflakeConfig.scala:
```diff
@@ -21,48 +21,50 @@ import scala._
+    maxTransactionReattempts getOrElse SnowflakeConfig.DefaultMaxReattempts
+
+  val jdbcUri: String =
+    s"jdbc:snowflake://${accountName}.snowflakecomputing.com/?db=${databaseName}&schema=${schema}&warehouse=${warehouse}&CLIENT_SESSION_KEEP_ALIVE=true&AUTOCOMMIT=false"
```
Can we pull the snowflake domain into a constant?

cryogenian replied:
Done

wemrysi commented on core/src/main/scala/quasar/destination/snowflake/StageFile.scala:
```diff
@@ -0,0 +1,59 @@
+object StageFile {
+  private val Compressed = true
+
+  def apply(input: Chunk[Byte], connection: SnowflakeConnection, blocker: Blocker)
```
Mind retaining the debug logging around pre-upload/post-delete?

cryogenian replied:
Done

wemrysi commented on core/src/main/scala/quasar/destination/snowflake/TempTableFlow.scala:
```diff
@@ -0,0 +1,287 @@
+        def ingest(stageFile: StageFile): ConnectionIO[Unit] = runFragment {
+          fr"COPY INTO" ++ tmpFragment ++ fr0" FROM @~/" ++
+          stageFile.fragment ++
+          fr""" file_format = (type = csv, skip_header = 0, field_optionally_enclosed_by = '"', escape = none)"""
```
Looking at the previous version, we're missing `escape_unenclosed_field = none`, does `escape = none` imply it, is it no longer needed?

cryogenian replied:
This param is needed, thanks for spotting this!

wemrysi commented on core/src/test/scala/quasar/destination/snowflake/QueryGenSpec.scala:
```diff
@@ -21,13 +21,13 @@ import org.specs2.mutable.Specification
   "SanitizeIdentifiers = false" >> {
     "quotes identifiers and removes double quotes" >> {
-      QueryGen.sanitizeIdentifier("includes \" spaces", SanitizeIdentifiers(false)) must_== """"includes  spaces""""
+      QueryGen.sanitizeIdentifier("includes \" spaces", false) must_== """"includes "" spaces""""
```
Mind updating the description to match the intent (quotes are escaped). Is this the desired behavior?

cryogenian replied:
Done :+1:

wemrysi commented on core/src/test/scala/quasar/destination/snowflake/QueryGenSpec.scala:
```diff
@@ -21,13 +21,13 @@ import org.specs2.mutable.Specification
   "SanitizeIdentifiers = true" >> {
     "quotes identifiers, makes everything uppercase and removes double quotes" >> {
-      QueryGen.sanitizeIdentifier("? foo>>bar123 ;;", SanitizeIdentifiers(true)) must_== "__FOO__BAR123___"
+      QueryGen.sanitizeIdentifier("? foo>>bar123 ;;", true) must_== "__FOO__BAR123___"
```
I realize it didn't have one before, but we could add a double quote to show it is removed (or modify the description).

cryogenian replied:
:+1:

cryogenian commented:
+ Removed `json` from code block in readme
+ Added `escape_unenclosed_field = none` to `COPY INTO`
+ `debug` in `StageFile` (also in releasing)
+ Texting in query gen tests

cryogenian commented on README.md:
```diff
@@ -15,7 +15,10 @@ libraryDependencies += "com.precog" %% "quasar-destination-snowflake" % <version
    user: String,
    password: String,
-   schema: String
+   schema: String,
```
Done

beckyconning commented:
Need this in a couple of hours so I'm going to go ahead and merge, feel free to continue the review and resolve process post merge.

wemrysi commented:
LGTM 👍
